iceoryx_hoofs/source/posix_wrapper/named_semaphore.cpp

```diff
@@ -0,0 +1,221 @@
+                                              convertToOflags(openMode),
+                                              static_cast<mode_t>(permissions),
+                                              static_cast<unsigned int>(initialValue))
+                      .failureReturnValue(IOX_SEM_FAILED)
+                      .evaluate();
+
+    if (result.has_error())
+    {
+        switch (result.get_error().errnum)
+        {
+        case EACCES:
+            LogError() << "Insufficient permissions to create semaphore \"" << name << "\".";
+            return cxx::error<SemaphoreError>(SemaphoreError::PERMISSION_DENIED);
+        case EEXIST:
+            LogError() << "A semaphore with the name \"" << name
+                       << "\" does already exist. This should not happen until there is a race condition when multiple "
+                          "instances try to create the same named semaphore concurrently.";
+            return cxx::error<SemaphoreError>(SemaphoreError::ALREADY_EXIST);
+        case EMFILE:
+            LogError() << "The per-process limit of file descriptor exceeded while creating the semaphore \"" << name
+                       << "\"";
+            return cxx::error<SemaphoreError>(SemaphoreError::FILE_DESCRIPTOR_LIMIT_REACHED);
+        case ENFILE:
+            LogError() << "The system wide limit of file descriptor exceeded while creating the semaphore \"" << name
+                       << "\"";
+            return cxx::error<SemaphoreError>(SemaphoreError::FILE_DESCRIPTOR_LIMIT_REACHED);
+        case ENOMEM:
+            LogError() << "Insufficient memory to create the semaphore \"" << name << "\".";
+            return cxx::error<SemaphoreError>(SemaphoreError::OUT_OF_MEMORY);
+        default:
+            LogError() << "This should never happen. An unknown error occurred while creating the semaphore \"" << name
+                       << "\".";
+            return cxx::error<SemaphoreError>(SemaphoreError::UNDEFINED);
+        }
+    }
+
+    constexpr bool HAS_OWNERSHIP = true;
+    uninitializedSemaphore.emplace(result.value().value, name, HAS_OWNERSHIP);
+
+    return cxx::success<>();
+}
+
+cxx::expected<SemaphoreError>
+NamedSemaphoreBuilder::create(cxx::optional<NamedSemaphore>& uninitializedSemaphore) noexcept
+{
+    if (!cxx::isValidFileName(m_name))
+    {
+        LogError() << "The name \"" << m_name << "\" is not a valid semaphore name.";
+        return cxx::error<SemaphoreError>(SemaphoreError::INVALID_NAME);
+    }
+
+    if (m_initialValue > IOX_SEM_VALUE_MAX)
+    {
+        LogError() << "The semaphores \"" << m_name << "\" initial value of " << m_initialValue
+                   << " exceeds the maximum semaphore value " << IOX_SEM_VALUE_MAX;
+        return cxx::error<SemaphoreError>(SemaphoreError::SEMAPHORE_OVERFLOW);
+    }
+
+    if (m_openMode == OpenMode::OPEN_EXISTING)
+    {
+        auto result = IOX_TRY(tryOpenExistingSemaphore(uninitializedSemaphore, m_name), result);
```

Comment:
Basically I'm not against such a macro but this implementation does not look quite ergonomic. It is even error prone.

Would you mind to remove it from this PR and introduce it in another one?

Reply:
no problem at all!